paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/Projection.java

```diff
@@ -184,6 +207,66 @@ public RowType project(RowType rowType) {
+                    new MutableRowType(rowType.isNullable(), Collections.emptyList());
+            for (int[] indexPath : this.projection) {
+                org.apache.paimon.types.RowType sourceType = rowType;
+                MutableRowType targetType = result;
+                int index;
+                for (index = 0; index < indexPath.length - 1; index++) {
+                    String fieldName = sourceType.getFieldNames().get(indexPath[index]);
+                    DataField field = sourceType.getField(fieldName);
+                    sourceType =
+                            (org.apache.paimon.types.RowType) sourceType.getField(fieldName).type();
+                    if (!targetType.containsField(fieldName)) {
+                        targetType.appendDataField(
+                                fieldName,
+                                field.id(),
+                                new MutableRowType(
+                                        sourceType.isNullable(), Collections.emptyList()),
+                                field.description());
+                    }
+                    targetType = (MutableRowType) targetType.getField(fieldName).type();
+                }
+
+                String fieldName = sourceType.getFieldNames().get(indexPath[index]);
+                DataField field = sourceType.getField(fieldName);
+                targetType.appendDataField(
+                        fieldName,
+                        field.id(),
+                        sourceType.getField(fieldName).type(),
+                        field.description());
+            }
+            return result.toRowType();
+        }
+
+        @Override
+        public ProjectionRowData getOuterProjectRow(org.apache.paimon.types.RowType rowType) {
+            org.apache.paimon.types.RowType resultType = project(rowType);
+
+            int[][] resultIndices = new int[this.projection.length][];
+            for (int i = 0; i < this.projection.length; i++) {
+                org.apache.paimon.types.RowType sourceType = rowType;
+                org.apache.paimon.types.RowType targetType = resultType;
+                resultIndices[i] = new int[this.projection[i].length];
+                for (int j = 0; j < this.projection[i].length; j++) {
+                    DataField sourceField = sourceType.getFields().get(this.projection[i][j]);
+                    String fieldName = sourceField.name();
+                    resultIndices[i][j] = targetType.getFieldIndex(fieldName);
+                    if (j < this.projection[i].length - 1) {
+                        targetType =
+                                (org.apache.paimon.types.RowType)
+                                        targetType.getField(fieldName).type();
+                        sourceType = (org.apache.paimon.types.RowType) sourceField.type();
+                    }
+                }
+            }
+
+            return new ProjectionRowData(toLogicalType(resultType), resultIndices);
```

Comment:
Why here must to return a row? I think if there is no nested projection, we don't need to do outer project again.